### PR TITLE
feat(desktop): open markdown attachments in an in-app viewer panel

### PR DESCRIPTION
**Pull request by rileycrane**
## Summary

Port of berd's markdown-link viewer into Buzz desktop. Clicking a `.md`/`.markdown`/`.mdx` attachment card now opens a right-side auxiliary panel rendering the document with the same markdown pipeline as chat messages, instead of forcing a download.

**What's in it**

- **`MarkdownDocPanel`** — right auxiliary panel with **Preview / Code** toggle, download button, and close. Hosted in `ChannelScreen` alongside threads/profiles; participates in narrow-layout single-panel mode and Esc-to-close like the other aux panels.
- **FileCard affordance** — markdown attachments show an open-panel icon; the card pre-gates on advertised size so a >2 MiB doc never double-fetches just to refuse.
- **Panel state in URL** (`doc`/`docName` params via `useChannelPanelHistoryState`) — back/forward and reload restore the open document.
- **Native 2 MiB cap** — the viewer fetches through a markdown-specific `fetch_markdown_doc_bytes` Tauri command whose cap is owned by Rust: oversized Content-Length is refused before the body is read, and the streamed byte count aborts mid-transfer when the header is missing or dishonest (a forged/absent imeta `size` can no longer buy a 50 MiB download + IPC copy through the generic media path). Both enforcement points are pure, unit-tested helpers. The command deliberately skips MIME sniffing — bytes are strictly UTF-8-decoded and rendered escaped, and a legitimate `.md` may begin with bytes that sniff as a blocked type.
- **`markdownDocFile.ts`** — strict UTF-8 decode + matching frontend cap; binary-or-oversized files fall back to a friendly error + Download.
- **`useChannelPaneOpeners`** — pane-open handlers extracted from `ChannelScreen`; upstream's `requireThreadEditResolution()` guard runs first in every opener (same contract as `useChannelProfilePanel`).
- **Nullable `MarkdownDocViewerProvider`** — surfaces without a panel host (forum/projects) keep plain download behavior.
- **Aux-pane split** (4th commit) — rebasing onto main pushed `ChannelPane.tsx`/`ChannelScreen.tsx` over the frozen file-size ratchet, so the agent-session and markdown-doc pane assemblies are extracted into `AgentSessionAuxiliaryPanel.tsx` + `MarkdownDocAuxiliaryPanel.tsx` (pure JSX moves, no behavior change; repo policy: split, never bump).
- **Selection-tray crash fix** (5th commit, pre-existing upstream bug hit while dogfooding this branch) — tiptap v3's `editor.view` accessor throws until EditorContent attaches the view, and `SelectionFormattingTray`'s wiring effect read `editor.view.dom` unguarded; navigating to a channel in a dev build could crash the whole channel route into TanStack Router's error screen. The tray now tracks view availability (probe + tiptap `mount`/`unmount` events) and gates its DOM wiring on it, with a regression test pinning the probe.

**Why it works differently from berd** — the Nostr/relay foundation changed three design points:

1. **Filename is the only identity.** The relay content-addresses uploads (`{sha256}.bin`, `application/octet-stream`) and markdown has no magic bytes — the ".md-ness" survives only in the NIP-92 imeta filename tag, so classification is filename-based.
2. **Authenticated fetch, not file reads.** Relay media URLs 401 in a plain browser (Blossom auth), so content goes through an authenticated Tauri command (SSRF-guarded to the relay `/media/` origin). This is also why in-app rendering is the only viable UX — "open in browser" can never work.
3. **Content-addressed = immutable** — doc content is cached with `staleTime: Infinity` per session.

No DOMPurify dance needed: buzz's markdown pipeline has no raw-HTML support (no rehype-raw), so doc HTML renders escaped and `javascript:` hrefs are stripped.

**Known limit** (documented in-code): Code view skips syntax highlighting past 150 lines — inherited shiki perf guard shared with chat code blocks.

### Related issue

N/A — none found (searched issues/PRs for "markdown viewer": no matches).

### Testing

At tip `aeffcce25`, based on main `f6e6617a9` (no file overlap with newer main commits):

- Tauri Rust suite green (`desktop-tauri-checks` lane), including 4 new cap tests: over-cap Content-Length refused, at-cap/absent admitted, streamed cutoff at exactly 2 MiB + 1 with nothing past the cap buffered; `cargo clippy -D warnings` + `cargo fmt --check` clean
- Desktop unit suite: 5442 passed / 0 failed (includes 8 `markdownDocFile` decoder tests + 2 tray view-guard tests)
- Playwright e2e: `markdown-doc-viewer.spec.ts` (upload → open panel → Preview/Code → download fallback → non-md regression → reload/back-forward contract → forged-size over-cap fallback) + `file-attachment.spec.ts` baseline: 21/21 passed
- `desktop-check` (biome/px-text/pubkey-truncation), `desktop-typecheck`, `file-size-check`: clean
- All 7 pre-push lanes green at push time

Screenshots: post via `scripts/post-screenshots.sh <PR#> ~/.buzz/.scratch/md-viewer` after the PR exists (repo policy — no relay media URLs in PR bodies).

### CODE REVIEW

**Comment by rileycrane** [comment]
### Preview view
A markdown attachment card (with the open-in-panel affordance) and the document rendered in the right auxiliary panel — headings, tables, task lists, and code through the same pipeline chat messages use.

md-viewer-preview

### Code view
The Preview / Code toggle showing raw source with syntax highlighting, plus the header download fallback.

md-viewer-code

**Comment by rileycrane** [comment]
## Screenshots

pr-6731--md-viewer-code

pr-6731--md-viewer-preview

File: desktop/src/features/messages/ui/SelectionFormattingTray.tsx [hunk 1]
```diff
@@ -25,6 +25,16 @@ function clamp(value: number, min: number, max: number): number {
   return Math.min(Math.max(value, min), max);
 }
 
+export function isEditorViewMounted(editor: Editor): boolean {
+  // The proxy behind `editor.view` throws before mount; probing is the only
+  // synchronous way to know whether the view is attached yet.
+  try {
+    return Boolean(editor.view);
```
Comment (wesbillman) [inline desktop/src/features/messages/ui/SelectionFormattingTray.tsx]: **[P1] Probe a property that actually establishes view availability.** In the installed Tiptap 3.22.5, `editor.view` itself never throws before mount: it returns a truthy Proxy. Only an unavailable property access such as `editor.view.dom` throws. I reproduced the real state after `editor.unmount()`: `Boolean(editor.view) === true`, followed by `editor.view.dom` throwing `The editor view is not available`. This helper therefore sets `viewMounted` true, and the wiring effect immediately repeats the channel-crashing read at line 195. The new test models a getter that throws and consequently passes without representing the dependency. Please key this off the actual mount lifecycle (or probe a property that proves attachment) and add a Proxy-faithful lifecycle regression covering pre-mount, unmount, remount, stale events, and cleanup.

File: desktop/src/features/channels/ui/MarkdownDocPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  url,
+  filename,
+  isSinglePanelView = false,
+  layout = "standalone",
+  onClose,
+  transparentChrome = false,
+  widthPx,
+}: MarkdownDocPanelProps) {
+  const isOverlay = useIsThreadPanelOverlay();
+  useEscapeKey(onClose, isOverlay || isSinglePanelView);
```
Comment (wesbillman) [inline desktop/src/features/channels/ui/MarkdownDocPanel.tsx:82]: **[P2] Move focus into this panel and restore it on close in the narrow layout.** Below 600 px, opening a document switches to single-panel mode and unmounts the channel section containing the focused attachment card. This panel registers Escape but neither focuses a panel control/landmark nor records/restores the opener. At 560×720, keyboard-opening the card leaves `document.activeElement === BODY`; after Escape, focus lands on the composer rather than the invoking attachment. Keyboard and screen-reader users therefore lose their position and cannot reach the panel controls without traversing the page again. Focus the close button or panel heading on open, restore the invoking card (or an intentional equivalent if it no longer exists) on close, and add narrow-layout focus assertions to the viewer E2E.

**Review by wesbillman — CHANGES_REQUESTED** [review]
Carl, an automated reviewer, commenting via Wes’s GitHub account.

Reviewed head `aeffcce256f20e30b45d92d0a8db8216916a31a8` against current base `30d2fc52f96138311f2006627ffc1a6d5ff1865b` (merge-base `f6e6617a9dcc2308d5039f8afaab974b49fb9577`). Requesting changes for one channel-crash regression and one narrow-layout accessibility defect.

I also traced the attachment producer/consumer surfaces, download fallbacks, pane lifecycle, native same-relay/auth boundary, redirect policy, streaming 2 MiB cap, strict UTF-8 decode, and safe markdown rendering. I found no additional blocking defect there. All current GitHub checks pass, but they do not exercise either behavior below faithfully.

**Comment by rileycrane** [comment]
Updated for the review-feedback commits (view picker moved to its own row below the title).

### Preview view
Markdown attachment card with the open-in-panel affordance; the document renders in the right auxiliary panel with the Preview/Code picker on its own row, so the filename keeps the title row.

md-viewer-preview

### Code view
Raw source with syntax highlighting; download fallback stays in the header.

md-viewer-code

File: desktop/src/features/messages/ui/SelectionFormattingTray.tsx (hunk 1 above)
Comment (rileycrane) [reply desktop/src/features/messages/ui/SelectionFormattingTray.tsx]: 🤖 (Fizz, on Riley's behalf) You're right — the pre-mount `editor.view` proxy is itself truthy and only a property read throws, so the truthiness probe reported mounted and re-ran the crashing read. Fixed in `0f30259e7`: the probe now reads `editor.view.dom` inside the try, the tracking is extracted into a shared `useEditorViewMounted` hook (probe at subscription time + tiptap's `mount`/`unmount` events), and the shallow throwing-getter test is replaced with a Proxy-faithful lifecycle suite — the fake's view is a truthy Proxy that throws on any property get, and the suite covers pre-mount, mount, unmount, remount, already-mounted-at-subscription, stale events from a replaced editor, and listener cleanup. The first assertion pins the trap itself: `Boolean(editor.view) === true` while `editor.view.dom` throws, so the old probe fails the new tests.

File: desktop/src/features/channels/ui/MarkdownDocPanel.tsx (hunk 2 above)
Comment (rileycrane) [reply desktop/src/features/channels/ui/MarkdownDocPanel.tsx:82]: 🤖 (Fizz, on Riley's behalf) Fixed in `c4175f4ee`. On mount the panel moves focus to its close control; on unmount it returns focus to the attachment card that opened the document, matched by a new `data-doc-url` anchor on FileCard since the original element was unmounted and remounted with the narrow layout swap. The restore intentionally overrides only free focus (`<body>`/null) and the remounted channel's composer autofocus — the exact 'lands on the composer' behavior you observed — and yields to any other control that claimed focus (e.g. a different panel). A narrow-layout (560×720) e2e now asserts keyboard-open lands focus on the panel close control and Escape returns it to the invoking card.

File: desktop/src/features/channels/ui/markdownDocFocus.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * After the panel closes, return focus to the attachment card that opened
+ * `url` once the channel section has remounted. Aborts if focus has already
+ * landed on a real control (e.g. a different panel claimed it), and gives up
+ * quietly when the card no longer exists.
+ */
+export function restoreFocusToMarkdownDocOpener(url: string): void {
+  scheduleFocusSearch(
+    () =>
+      document.querySelector<HTMLElement>(
+        `[data-testid="file-card"][data-doc-url="${CSS.escape(url)}"]`,
```
Comment (wesbillman) [inline desktop/src/features/channels/ui/markdownDocFocus.ts]: **[P2] Preserve the invoking card when document URLs repeat.** This selector identifies the opener only by URL. If the same relay attachment appears in two messages, both cards have the same `data-doc-url`, and `querySelector` always returns the first DOM match. I reproduced this at 560×720 with two markdown cards sharing one URL: keyboard-open the second card, press Escape, and focus restores to the first card instead of the invoking second card. Carry a stable card/message identity or a unique restoration token through the panel state (or an equivalent per-invocation anchor), and add the duplicate-URL narrow-layout regression. The URL alone cannot identify the opener instance.

**Review by wesbillman — CHANGES_REQUESTED** [review]
Carl, an automated reviewer, commenting via Wes’s GitHub account.

Reviewed the follow-up at `c4175f4ee50733b071d6123c13c956605308ea44`. The Tiptap lifecycle correction is sound: the new probe exercises `.dom` on the real pre-mount truthy proxy, mount/unmount events match actual view availability, replacement and listener cleanup are covered, and the targeted narrow-layout focus test passes.

The focus-restoration follow-up still has one blocking P2 edge case: a document URL is not a unique opener identity. I reproduced the duplicate-attachment case below in the E2E app; opening the second card and closing the panel focuses the first card. Requesting a stable per-invocation restoration identity and regression coverage.

File: desktop/src/features/channels/ui/markdownDocFocus.ts (hunk 3 above)
Comment (rileycrane) [reply desktop/src/features/channels/ui/markdownDocFocus.ts]: 🤖 (Fizz, on Riley's behalf) Fixed in `6f11ce55a`. `FileCard` now passes the clicked element through the open call, and the pane opener records it as its DOM-order index among the cards sharing that URL — captured at open time while the card is still mounted, so the identity survives the narrow-layout unmount/remount that invalidates an element reference. The restore prefers the recorded index (clamped if cards disappeared), consumes the record so a deep-link/reload open cannot inherit a stale index, and keeps the existing yield-to-claimed-focus behavior. Coverage: a duplicate-URL 560×720 keyboard e2e (open the second of two same-URL cards, Escape, assert the second — not the first — is focused) plus a JSDOM unit suite pinning the record/restore contract (recorded card wins, consumption, clamping, URL mismatch ignored, claimed-focus abort).

File: desktop/src/features/messages/ui/SelectionFormattingTray.tsx (hunk 1 above)
Comment (rileycrane) [reply desktop/src/features/messages/ui/SelectionFormattingTray.tsx]: 🤖 (Fizz, on Riley's behalf) Update after rebasing onto latest main: #6779 landed upstream with its own fix for this same pre-mount race (`getMountedEditorDom` probe + mount/unmount listener reattachment, staging-verified). To avoid carrying a competing mechanism for an already-fixed upstream bug, this branch's two tray commits (`useEditorViewMounted` gate + tests) were dropped during the rebase — the PR no longer touches `features/messages` at all and inherits main's fix. Current head `7aae92ba5`.

File: desktop/src/features/channels/ui/channelPaneAuxiliaryLayout.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  const useFocusIdleDrawer = shouldUseFocusIdleDrawer({
+    channelManagementOpen,
+    hasAgentSession,
+    hasIdleAuxiliaryPanel,
+    hasIdlePanelCloseHandler,
+    hasProfilePanel,
+    hasThreadSurface,
+    overrideThread: overlayIdleAuxiliaryOverThread,
+    useSplitAuxiliaryPane,
+  });
```
Comment (wesbillman) [inline desktop/src/features/channels/ui/channelPaneAuxiliaryLayout.ts:63]: **[P2] Arbitrate the document pane before activating idle-drawer coverage.** This calculation does not consider an open markdown document, while `ChannelPane` renders the document ahead of `idleAuxiliarySurface` (lines 977–989). The combination is reachable through the real project-home controls: open a `.md` attachment in the project conversation, then click **Tasks** in the project context. At this head, the document remains visible, neither the workspace sheet nor a focus drawer renders, but `useFocusIdleDrawer` makes `channel-drop-zone` inert, including the message composer. I reproduced that sequence in the current-head Chromium E2E app. Resolve the active pane consistently for rendering and coverage (or close/suspend the document when opening the workspace sheet); an invisible drawer must not disable the conversation. Add a project-home document → Tasks regression asserting the selected sheet is visible and inertness belongs to an actually rendered modal surface.

File: desktop/src/features/channels/ui/markdownDocFocus.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+export function restoreFocusToMarkdownDocOpener(url: string): void {
+  scheduleFocusSearch(
+    () => {
+      const cards = findOpenerCards(url);
+      if (cards.length === 0) return null;
+      const record = lastOpenerRecord?.url === url ? lastOpenerRecord : null;
+      // One record per invocation: consume it so a later open without a
+      // recorded card (deep link, reload) does not inherit a stale index.
+      if (record) lastOpenerRecord = null;
+      return cards[Math.min(record?.index ?? 0, cards.length - 1)] ?? null;
```
Comment (wesbillman) [inline desktop/src/features/channels/ui/markdownDocFocus.ts]: **[P2] Restore by stable attachment identity, not a mutable list position.** Recording an ordinal fixes the static two-card case but not a changing timeline. Reproduction at 560×720: render three same-URL attachments in messages A/B/C, keyboard-open B, receive a live kind:5 deletion of A, then Escape. B survives, but this lookup focuses C because B shifted from index 1 to index 0. I independently reproduced it through the E2E live-event bridge: surviving IDs were B/C, and the active card belonged to C. Carry a stable message/card restoration token across the narrow-layout remount and use a fallback only when the exact opener is gone. The current disappearance unit case removes the opener itself, so it does not cover deletion of a preceding sibling.

File: desktop/src/features/channels/ui/useChannelPaneOpeners.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      // Opening a doc closes the thread pane, so an in-progress thread edit
+      // must be resolved first — same contract as the sibling pane openers.
+      if (!requireThreadEditResolution()) return;
+      // Capture the invoking card's identity before the panel swap unmounts
+      // it, so close can restore focus to this card and not merely the first
+      // card sharing the document URL.
+      recordMarkdownDocOpener(doc.url, doc.opener ?? null);
+      clearCompetingPanes();
+      setChannelManagementOpen(false);
+      openMarkdownDoc(doc.url, doc.filename);
```
Comment (wesbillman) [inline desktop/src/features/channels/ui/useChannelPaneOpeners.ts:103]: **[P2] Preserve a logical return-focus target for thread-origin documents.** `clearCompetingPanes()` clears the originating thread, but the recorded opener contains only a URL/index, and close only removes `doc`/`docName`. For a markdown attachment in a thread reply with no main-timeline copy, closing the viewer leaves no matching card for `restoreFocusToMarkdownDocOpener()` to find. I independently reproduced opening the reply's card and keyboard-activating Close: after restoration/exit frames, the thread is still closed and focus is BODY at 1280px or the main composer at 560px. Keyboard/screen-reader users lose their position on this ordinary opener surface, even without duplicates or deletions. Preserve enough origin context to restore the thread/card, or deliberately focus its surviving thread-summary control; add wide and narrow thread-origin close assertions. This is separate from the unstable-index case because a stable card token alone cannot restore a card whose surface never remounts.

**Review by wesbillman — CHANGES_REQUESTED** [review]
Carl, an automated reviewer, commenting via Wes’s GitHub account.

Reviewed head `263e1aeed116d9bd02839ecbea7289c335f188bb` against base `0ccf934b88f610f5f235862ecd51e2dcbae2cb74`. Requesting changes for three reproduced P2 UI/accessibility defects below.

The original static duplicate-card and narrow-layout focus cases now pass, and the branch inherits main's Tiptap lifecycle fix. The remaining issues are dynamic opener identity, thread-origin return focus, and project-home auxiliary-pane arbitration.

Validation: source/contract review plus focused Chromium workflows against a fresh current-head E2E build with the mock Tauri/live-event bridge. I independently reran the live-deletion and thread-origin reproductions and reproduced the project-home conflict. This is not native WKWebView or live-relay validation. Focused decoder/attachment tests reported 91 passes; a separate markdown renderer test file could not load because the shared review checkout lacks `linkifyjs`. No fresh CI or full-suite/native pass is claimed.

**Comment by wesbillman** [comment]
Screenshot 2026-08-29 at 7 30 16 AM

The icon and title here seem a bit misaligned.

**Comment by delkc** [comment]
The markdown files aren't getting this functionality from the Inbox view, but seems like a pretty easy fix (my agent patched it locally for me). Ideally it would be great if we can have it work from any view.

Having the file fully replace a thread in that right panel when you open from there is a little jarring, so I wonder if it could either push the thread into the center like it does when you hit the 3 dots overflow menu in the top right OR if it could pop out over view the view like it does when you hit the "Expand thread" action in the top left of the panel?

**Comment by delkc** [comment]
🤖 I tested this branch against the team relay and put the follow-up fixes into one atomic commit for consideration:

**Suggested commit:** https://github.com/block/buzz/commit/032d2bd4d

```bash
git cherry-pick 032d2bd4d
```

### What this changes

- **Makes the Preview action stable.** Removes the renderer-side relay-origin timing gate that caused valid Markdown cards to alternate between Preview and Download. The native fetch command remains the authoritative relay-origin and 2 MiB security boundary.
- **Handles where the file was opened.** A document opened from the center timeline uses the normal right panel. Opening from an existing thread preserves that mounted thread rather than replacing it.
- **Adds responsive thread behavior.** When the channel container can fit two saved panel widths plus a 300px conversation column, the thread shifts left and the document opens as a third panel. Below that threshold, the document stacks over the thread at the existing right-panel width.
- **Keeps Inbox local.** Opening a document from Inbox no longer navigates into the channel route. On wide windows it becomes a third Inbox panel; on constrained windows it stacks over the Inbox detail. Closing restores the same Inbox conversation.
- **Preserves focus and state.** Closing returns to the invoking attachment where possible, while the underlying thread/Inbox detail remains mounted with its scroll and reply state intact.
- **Adds regression coverage** for thread-reply attachments and responsive pane arbitration.

### Verification

- TypeScript typecheck
- Repository file-size gate
- 52 focused unit tests
- Biome / `git diff --check`
- Manual team-relay testing across timeline, thread, DM, and Inbox contexts

I used a commit rather than inline suggestion blocks because the change adds several coordinated files and needs to apply atomically.
